lib/keystore.cc

```diff
@@ -82,55 +130,113 @@ rpmRC keystore_fs::delete_key(rpmtxn txn, rpmPubkey key)
 rpmRC keystore_fs::import_key(rpmtxn txn, rpmPubkey key, int replace, rpmFlags flags)
 {
     rpmRC rc = RPMRC_FAIL;
-    const char *fp = rpmPubkeyFingerprintAsHex(key);
-    char *keyfmt = rstrscat(NULL, "gpg-pubkey-", fp, ".key", NULL);
-    char *keyval = rpmPubkeyArmorWrap(key);
-    char *path = rpmGenPath(rpmtxnRootDir(txn), "%{_keyringpath}/", keyfmt);
-    char *tmppath = NULL;
+    string fp = rpmPubkeyFingerprintAsHex(key);
+    string keyfmt = "gpg-pubkey-" + fp + ".key";
+    char *dir = rpmGenPath(rpmtxnRootDir(txn), "%{_keyringpath}/", NULL);
+    string dirstr = dir;
 
-    if (replace) {
-	rasprintf(&tmppath, "%s.new", path);
-	unlink(tmppath);
+    rc = write_key_to_disk(key, dirstr, keyfmt, replace, flags);
+
+    if (!rc && replace) {
+	/* find and delete the old pubkey entry */
```

Comment:
I'm a bit concerned about this.  OpenPGP certificates are mostly append-only data structures. To understand why, imagine that the certificate is revoked, and the revocation certificate is stored locally.  Then some process comes along and updates the certificates, but it doesn't have the revocation certificate for some reason.  The revocation certificate will now be deleted, and rpm may start relying on the revoked certificate again.

So, from my perspective, the correct thing to do here is to merge the existing data with the new data.

Can you explain your thought process here and what you are trying to accomplish?

Reply:
The patch obscures that this is actually old code. This is about converting from the old (short keyids based) file names to new (fingerprint based) filenames. Here we are just deleting the old file which is now redundant.

The merging of the keys does happen - it just doesn't happen here.

Reply:
Yeah, I thought it wouldn't make a big difference having it in one patch as it just adds the new backend. I should have reconsidered when seeing the mess the diff made out of that.